src/glang/lib/std/unicode.c3

```diff
@@ -0,0 +1,412 @@
+        buffer:get(1) = Narrow<u8>(0x80 | (scalar.value & 0x3f));
+        return {&buffer.data, 2};
+    }
+
+    if is_in_range(scalar.value, 0x0800, 0x0fff)
+        or is_in_range(scalar.value, 0x1000, 0xcfff)
+        or is_in_range(scalar.value, 0xd000, 0xd7ff)
+        or is_in_range(scalar.value, 0xe000, 0xffff)
+    {
+        buffer:get(0) = Narrow<u8>(0xe0 | ((scalar.value >> 12) & 0x0f));
+        buffer:get(1) = Narrow<u8>(0x80 | ((scalar.value >> 6) & 0x3f));
+        buffer:get(2) = Narrow<u8>(0x80 | (scalar.value & 0x3f));
+        return {&buffer.data, 3};
+    }
+
+    if is_in_range(scalar.value, 0x10000, 0x3ffff)
+        or is_in_range(scalar.value, 0x40000, 0xfffff)
+        or is_in_range(scalar.value, 0x100000, 0x10ffff)
+    {
+        buffer:get(0) = Narrow<u8>(0xf0 | ((scalar.value >> 18) & 0x07));
+        buffer:get(1) = Narrow<u8>(0x80 | ((scalar.value >> 12) & 0x3f));
+        buffer:get(2) = Narrow<u8>(0x80 | ((scalar.value >> 6) & 0x3f));
+        buffer:get(3) = Narrow<u8>(0x80 | (scalar.value & 0x3f));
+        return {&buffer.data, 4};
+    }
+
+    // Graphene cannot generate an invalid unicode scalar via `make<>` or the iterators
+    // This is a bug in the user's code so I don't mind the unrecoverable error here
+    runtime_assert(false);
+    return {&buffer.data, 0};
+}
+
+function encode_to_utf8 : (scalar : UnicodeScalar, buffer: u8[4] mut&) -> View<u8> = {
+    return encode_to_utf8(scalar, make<Span<u8>>(&mut buffer));
+}
+
+// --------- UTF16 -----------
+function get_next_codepoint : (self: UTF16DecodeIter mut&) -> Optional<u16> = {
+    if !self.underlying:has_next() { return make<Optional<u16>>(); }
+    let first_byte : u8 = self.underlying:__builtin_get_next();
+
+    if !self.underlying:has_next() { return make<Optional<u16>>(); }
+    let second_byte : u8 = self.underlying:__builtin_get_next();
+
+    if self.is_little_endian {
+        return make<Optional<u16>>(first_byte | (as_u16(second_byte) << Narrow<i16>(8)));
+    } else {
+        return make<Optional<u16>>(second_byte | (as_u16(first_byte) << Narrow<i16>(8)));
+    }
+}
+
+@implicit get_next : (self: UTF16DecodeIter mut&) -> UnicodeScalar = {
+    let maybe_codepoint1 : Optional<u16> = self:get_next_codepoint();
+    if !maybe_codepoint1:has_value() {
+        return invalid_unicode_scalar();
+    }
+
+    let codepoint1 : u16 = maybe_codepoint1:data();
+
+    // Is this the second part of a surrogate pair?
+    if codepoint1 & 0xfc00 == 0xdc00 {
+        return invalid_unicode_scalar();
+    }
+
+    // Is this a standalone codepoint?
+    if codepoint1 & 0xfc00 != 0xd800 {
+        return { codepoint1 };
+    }
+
+    // None of above, it must be the first part of a surrogate pair
+    let maybe_codepoint2 : Optional<u16> = self:get_next_codepoint();
+    if !maybe_codepoint2:has_value() {
+        return invalid_unicode_scalar();
+    }
+
+    let codepoint2 : u16 = maybe_codepoint2:data();
```

Comment:
Should we not check that this matches `110111xxxxxxxxxx`?

Reply:
Yes. Good spot.